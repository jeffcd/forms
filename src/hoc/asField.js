import React from 'react'
import get from 'lodash/get'
import FormContext from '../FormContext'
import VisibilityContext from '../VisibilityContext'
import ListContext from '../ListContext'
import validateField, { validatorFunctions } from '../validateField'
import messages from '../messages'
const validators = validatorFunctions

const getValidators = props => {
  const fieldValidators = {}
  Object.keys(props).forEach(name => {
    if (validators[name]) {
      fieldValidators[name] = props[name]
    }
  })
  return fieldValidators
}

const errorMessages = messages

const getErrorMessages = ({ props, field }) => {
  return field.errors.map(error => {
    let message = errorMessages[error]
    if (props.messages && props.messages[error]) {
      message = props.messages[error]
    }
    return message
      .replace('%name%', props.label)
      .replace('%param%', props[error])
  })
}

const getInitialValue = (Field, { minLength = 0 }) => {
  if (Field.type === 'list') {
    const size = parseInt(minLength)
    return Array(size)
  }

  return ''
}

const asFieldHoc = Field => {
  return function asFieldHoc({ ...rest }) {
    return (
      <ListContext.Consumer>
        {listInfo => {
          const { name, initialValue, convertTo } = rest
          const fullName = listInfo
            ? `${listInfo.name}.value[${listInfo.i}].${name}`
            : name

          const listProps = {}
          if (listInfo) {
            const { i, name } = listInfo
            const fullName = `${name}.value[${i}].${rest.name}`
            const id = `${name}:${i}::${rest.name}:`
            listProps.id = id
            if (rest.name) {
              listProps.name = fullName
            }
          }

          return (
            <FormContext.Consumer>
              {form => {
                //useEffect(() => {
                //}, [form.setFormField])

                const field = get(form.fields, fullName)
                if (!field) {
                  const field = {
                    name,
                    value: initialValue || getInitialValue(Field, rest),
                    validators: getValidators(rest),
                    convertTo,
                    errors: []
                  }
                  const isUser = false
                  form.setFormField(field, fullName, isUser)

                  return null
                }

                return (
                  <VisibilityContext.Consumer>
                    {visibility => {
                      if (visibility.isVisible === false) {
                        if (field.isVisible !== false) {
                          const isUser = false
                          field.isVisible = false
                          form.setFormField({ ...field }, isUser)
                        }
                        return null
                      }

                      const handleChange = e => {
                        const field = get(form.fields, fullName)
                        const newField = {
                          ...field,
                          value: e.target.value
                        }
                        const errors = validateField(newField)
                        newField.errors = errors
                        form.setFormField(newField, fullName)
                      }
                      return (
                        <Field
                          onChange={handleChange}
                          value={field.value}
                          errors={getErrorMessages({ props: rest, field })}
                          {...rest}
                          {...listProps}
                        />
                      )
                    }}
                  </VisibilityContext.Consumer>
                )
              }}
            </FormContext.Consumer>
          )
        }}
      </ListContext.Consumer>
    )
  }
}

export default asFieldHoc
