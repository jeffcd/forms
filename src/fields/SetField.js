import React, { useEffect } from 'react'
import asField from '../hoc/asField'

const SetField = asField(({ onChange, setValue }) => {
  useEffect(() => {
    onChange({ target: { value: setValue } })
  }, [setValue])
  return <></>
})

export default SetField
