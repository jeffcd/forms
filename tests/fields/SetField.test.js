import React from 'react'
import Form from '../../src'
import asField from '../../src/hoc/asField'
import SetField from '../../src/fields/SetField'

const Input = asField(
  ({
    onChange,
    onBlur = () => {},
    name,
    id,
    value,
    errors,
    label,
    helperText,
    required = false
  }) => {
    return (
      <div>
        <label htmlFor={name}>{label}</label>
        <br />
        <input
          id={`input-${id}`}
          name={name}
          value={value}
          onChange={e => onChange(e)}
          onBlur={e => onBlur(e)}
          label={label}
          required={required}
        />
        <span>{errors.length ? errors[0] : helperText}</span>
      </div>
    )
  }
)

it('Form: Set Field', () => {
  const submitSpy = jest.fn()
  let bodySpy = {}
  const form = (
    <Form
      onSubmit={(e, body, actions) => {
        bodySpy = body
        submitSpy()
      }}
    >
      <Input id="firstName" name="firstName" required />
      <SetField name="firstName" setValue="Jeff" />
      <SetField name="firstName" setValue="Jeff" />
      <button type="submit" id="submit-button">
        Submit
      </button>
    </Form>
  )
  const tree = mount(form)

  expect(tree.html()).toMatchSnapshot()

  tree.find('#submit-button').simulate('click')
  tree.find('form').simulate('submit')
  expect(tree.html()).toMatchSnapshot()
  expect(bodySpy).toMatchSnapshot()
  expect(submitSpy).toHaveBeenCalledTimes(1)
})
