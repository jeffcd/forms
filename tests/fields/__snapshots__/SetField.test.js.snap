// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form: Set Field 1`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"Jeff\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Set Field 2`] = `"<form novalidate=\\"\\" autocomplete=\\"off\\"><div><label for=\\"firstName\\"></label><br><input id=\\"input-firstName\\" name=\\"firstName\\" required=\\"\\" value=\\"Jeff\\"><span></span></div><button type=\\"submit\\" id=\\"submit-button\\">Submit</button></form>"`;

exports[`Form: Set Field 3`] = `
Object {
  "firstName": "Jeff",
}
`;
